src/Analyzers/CSharp/Analyzers/CodeStyle/CSharpAnalyzerOptionsProvider.cs

```diff
@@ -0,0 +1,142 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.CodeAnalysis.AddImport;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.CodeGeneration;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Formatting;
+using Microsoft.CodeAnalysis.CSharp.Simplification;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Options;
+
+namespace Microsoft.CodeAnalysis.Diagnostics;
+
+/// <summary>
+/// Provides C# analyzers a convenient access to editorconfig options with fallback to IDE default values.
+/// </summary>
+internal readonly struct CSharpAnalyzerOptionsProvider
+{
+    /// <summary>
+    /// Document editorconfig options.
+    /// </summary>
+    private readonly AnalyzerConfigOptions _options;
+
+    /// <summary>
+    /// Fallback options - the default options in Code Style layer.
+    /// </summary>
+    private readonly IdeAnalyzerOptions _fallbackOptions;
```

Comment:
A couple of thoughts: 
1) You can think of fallback options as any other feature options. If we didn't have editorconfig files the features would take these options as parameters like they take other options that are not stored in editorconfig files. We'd just call the parameters "options" instead of "fallbackOptions". They are only called "fallbackOptions" to emphasize that the feature should first look at document for editorconfig and only if the option is not found there it falls back to these options.
2) The idea behind the whole options refactoring is to make solution snapshot reflect the content of files on disk (in the repo). Options that are stored in registry are exogenous to the repo state. Storing them in the solution snapshot makes their management complex.